File: longestZigZag.java
import java.util.*;
// Approach 1- DFS 
class Solution { // TC O(n) SC O(n)
    int pathLength = 0;
    private void dfs(TreeNode node, boolean goLeft, int steps) {
        if (node == null) {
            return;
        }
        pathLength = Math.max(pathLength, steps);
        if (goLeft) {
            dfs(node.left, false, steps + 1);
            dfs(node.right, true, 1);
        } else {
            dfs(node.left, false, 1);
            dfs(node.right, true, steps + 1);
        }
    }

    public int longestZigZag(TreeNode root) {
        dfs(root, false, 0);
        dfs(root, true, 0);
        return pathLength;
    }
}

class longestZigZag{
    public static void main(String[] args) {
        BinaryTree bt=new BinaryTree();
        bt.root=new Node();
        Solution s= new Solution();
    }
}

File: longestZigZag.py
class Solution:
# Approach 1- DFS
    def longestZigZag(self, root): # TC O(n) // SC O(n)
        visited=0
        
        def dfs(root, isLeft, steps):
            nonlocal visited 
            
            if not root: return 
            
            visited=max(visited, steps)
            if isLeft: 
                dfs(root.left, False, steps+1)
                dfs(root.right, True, 1)
            else:
                dfs(root.left, False, 1)
                dfs(root.right, True, steps+1)
            
        dfs(root, False, 0)
        dfs(root, True, 0)
        return visited

s=Solution()
print(s.longestZigZag(root))
